scripts/systemd/xcash-dpops.service

```diff
@@ -1,13 +1,15 @@
 [Unit]
 Description=X-Cash DPOPS Program background process
+After=network.target xcash-rpc-wallet.service MongoDB.service
  
 [Service]
 Type=simple
 LimitNOFILE=infinity
 User=root
 WorkingDirectory=/root/xcash-official/xcash-dpops/build
+ExecStartPre=/usr/bin/sleep 10
```

Comment:
Is there a way we can not include this, since If someone needs to start them manually without the script, it would cause a 10 second delay? can we move the 10 second delay to inside the script instead?

Reply:
The same as the other sleep, but in this case would be easier to do it since all the code is your child and you know exactly where to put the "blucle" :)

Reply:
Here is not as easy as the rpc-wallet unit file since the dpops service exists if can't find the public address (because is still syncing or something)

```
Aug 11 15:54:41 x-node-1 xcash-dpops[1260]: Function Calls:
Aug 11 15:54:41 x-node-1 xcash-dpops[1260]: #1 get_public_address: Could not get the public address
Aug 11 15:54:41 x-node-1 xcash-dpops[1260]: #2 get_delegates_data: Could not get the wallets public address
```

I think here would be interesting to do some sort of bucle or something to just dont die if it can't get the address, but retry every 5-10 seconds for example until it can reach it

Reply:
Another way to do it without the sleep when you do `systemctl restart service` is to make a timer for `xcash-rpc-wallet` and `xcash-dpops`, for example:

```bash
#xcash-rpc-wallet.timer 
[Unit]
Description=xcash-rpc-wallet timer

[Timer]
OnBootSec=20sec

[Install]
WantedBy=timers.target
```
and 
```bash
#xcash-dpops.timer 
[Unit]
Description=xcash-dpops timer

[Timer]
OnBootSec=30sec

[Install]
WantedBy=timers.target
```

Reply:
What about adding a post delay to MongoDB service? And removing the pre-delay in DPOPS service
```
ExecStartPost=/usr/bin/sleep 20
```
In this case when you boot the system you have the delay before DPOPS start. If you restart only the DPOPS services manually or with the script without touching the mongodb there is no delay (you control the delay inside the script in this case). You have the delay only when starting mongodb the first time.

Also adding this to the DPOPS service
```
Restart=always
RestartSec=5
```
will restart the service also when it can not get the public address and without the "start-limit-hit" problem.
Is it a problem to always restart the DPOPS service when it exits due to public address error? It's not clean, but the trick works.
I noticed also a thing: if you run DPOPS and then stop the rpc-wallet service, the DPOPS service keeps going trying to connect to nodes. Is this the correct behavior or it should check if the wallet is alive? Or maybe is this check is done after the trying-to-connect loop?

Reply:
Hmm, the start post can work too, just need to make all xcash services to require the mongodb service, but with the timers I think its clear too, I have now the timers and is working fine without the "ugly" sleep

Reply:
MongoDB is needed only by DPOPS. So at boot mongoDB and xcash-daemon starts more or less the same time, then the RPC wallets waits for the daemon (with the RestartSec there can be a delay, but if everything is more or less in sync this will be less than 20s). And finally the DPOPS starts (that depends on mongoDB, so after 20s). If there are problems (sync longer...) it starts restarting every 5 sec (with RestartSec).

If you restart all services manually it works (maybe after some retries). If you restart with the autoinstaller everything is ok because of the delays inside the installer. So it should be fine for every case. Usually there are no problems with the mongo service, so the delay of 20s is only at boot or after the first install of the node.

Reply:
Sleeps are ugly, but with timers you need to create some extra files and the timers are bypassed if you use "systemctl enable" if I remember correctly. So you have to disable the services and use the timers only for those two. With the daemon you must still use "enable" and also with the mongo service...

Reply:
If you want you can omit the ExecStartPost, in this case you use only the auto-restart delay feature (RestartSec). Maybe it takes more rounds but eventually everything start... The ExecStartPost is only used to add some static delay without creating extra config files and remember to use "systemctl enable" only with daemon and mongodb services. 
The goal is the same... But you have to change also the documentation if timers are used and change the autoinstaller to write also these two files.

Reply:
> Sleeps are ugly, but with timers you need to create some extra files and the timers are bypassed if you use "systemctl enable" if I remember correctly. So you have to disable the services and use the timers only for those two (enable the timer, disable the related service). With the daemon you must still use "enable" (the service) and also with the mongo service...

Yes, you need to disable the service file and enable the timers but this way you can make the services delay the time you need and only for the first boot, after that, you can restart the services without any issue and without delay because you can make them restart in order with the sleeps (or including some bucle check until something happens) from the script itself.

Btw, keep in mind the timers can be generated automatically like the service ones and the user can just restart or take the needed actions with the service file, they will dont need to touch or even know there is a timer to control the startup.

Reply:
So the plan is to use 2 timers, or better to use 3 timers so the all the 3 xcash services are controlled by timers at boot (and have consistency)? So the autoinstaller creates the files and for enabling/disabling the boot feature you simply enable/disable the corresponding timer.
And I would keep the `RestartSec=5` as fallback if something goes wrong, both in rpc-wallet and dpops.

Reply:
This is the idea, but lets see what @zachhildreth think about all this :joy:

Reply:
Hi,
This sounds like a good solution to me, since we can leave the manual way of running it alone with no delays for when I am restarting them for testing. We can add delays to the script like we always have, but we can also include a "start at boot" feature for users that want that kind of setup.

Personally I am not interested in the auto boot feature for my delegate, but I think its a good idea especially if the community members want this feature.

**If someone can add to the repo the**
autoinstaller embedded timers files
code to create those timers where the other code is to create the service files
the timer files templates in the scripts/systemd folder
an question at install if you want auto boot or not (and the systemctl enable after installing timers if so)
a menu entry for enable/disable auto boot feature at the main menu (so you can switch it on/off) (and corresponding code to do it)
anything else I am forgetting to add this feature in?

Not sure if this would be better as a separate PR or included in this one, so I am good with either option

Thanks @rradu92 and @n3me5is-git

Reply:
I'm testing the autoinstaller with the timers now

Reply:
Awesome, thank you!

Reply:
Ok, finished the test and seems to work fine if you answer `yes|YES|ANY NON NULL STRING` to the `Do you want to autostart the services when you restart the server? (leave empty for default: NO): yes` 

Fixed too some situations when `RAM_CPU_RATIO` variable becomes 0 in low resources machines and breaks the  make command: `make release -j 0`

Reply:
Nice, Thanks @rradu92
I will give a test and pull it in in the next few days before we start beta 2